core/src/main/java/org/apache/iceberg/DefaultValueParser.java

```diff
@@ -0,0 +1,307 @@
+        }
+
+        return mapBuilder.build();
+      case STRUCT:
+        Preconditions.checkArgument(defaultValue.isObject(),
+            "Cannot parse default as a %s value: %s", type, defaultValue);
+        Types.StructType struct = type.asStructType();
+        StructLike defaultRecord = GenericRecord.create(struct);
+
+        List<Types.NestedField> fields = struct.fields();
+        for (int pos = 0; pos < fields.size(); pos += 1) {
+          Types.NestedField field = fields.get(pos);
+          String idString = String.valueOf(field.fieldId());
+          if (defaultValue.has(idString)) {
+            defaultRecord.set(pos, fromJson(field.type(), defaultValue.get(idString)));
+          }
+        }
+        return defaultRecord;
+      default:
+        throw new IllegalArgumentException(String.format("Type: %s is not supported", type));
+    }
+  }
+
+  public static Object fromJson(Type type, String defaultValue) {
+    try {
+      JsonNode defaultValueJN = JsonUtil.mapper().readTree(defaultValue);
+      return fromJson(type, defaultValueJN);
+    } catch (IOException e) {
+      throw new IllegalArgumentException("Failed to parse: " + defaultValue + "; reason: " + e.getMessage(), e);
+    }
+  }
+
+  public static String toJson(Type type, Object defaultValue) {
+    return toJson(type, defaultValue, false);
+  }
+
+  public static String toJson(Type type, Object defaultValue, boolean pretty) {
+    try {
+      StringWriter writer = new StringWriter();
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      if (pretty) {
+        generator.useDefaultPrettyPrinter();
+      }
+      toJson(type, defaultValue, generator);
+      generator.flush();
+      return writer.toString();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public static void toJson(Type type, Object javaDefaultValue, JsonGenerator generator)
+      throws IOException {
+    switch (type.typeId()) {
+      case BOOLEAN:
+        generator.writeBoolean((boolean) javaDefaultValue);
+        break;
+      case INTEGER:
+        generator.writeNumber((int) javaDefaultValue);
+        break;
+      case LONG:
+        generator.writeNumber((long) javaDefaultValue);
+        break;
+      case FLOAT:
+        generator.writeNumber((float) javaDefaultValue);
+        break;
+      case DOUBLE:
+        generator.writeNumber((double) javaDefaultValue);
+        break;
+      case DATE:
+        generator.writeString(DateTimeUtil.formatEpochDays((int) javaDefaultValue));
+        break;
+      case TIME:
+        generator.writeString(DateTimeUtil.formatTimeOfDayMicros((long) javaDefaultValue));
+        break;
+      case TIMESTAMP:
+        generator.writeString(DateTimeUtil.formatEpochTimeMicros(
+            (long) javaDefaultValue,
+            ((Types.TimestampType) type).shouldAdjustToUTC()));
+        break;
+      case STRING:
+        generator.writeString((String) javaDefaultValue);
```

Comment:
This is only required to be `CharSequence`, so you should cast to `CharSequence` and call `toString`.